src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java

```diff
@@ -0,0 +1,408 @@
+                        if (targetDataSource != null) {
+                            SAMSequenceDictionary targetDictionary = targetDataSource.getSequenceDictionary();
+                            SAMSequenceRecord targetSequence = targetDictionary.getSequence(sourceName);
+                            validateSequenceDictionaryRecords(
+                                    ds.getName(),
+                                    ds.getSequenceDictionary(),
+                                    sourceSequence,
+                                    targetDataSource.getName(),
+                                    targetDictionary,
+                                    targetSequence);
+                        } else {
+                            contigMap.put(sourceName, ds);
+                        }
+                    }
+                );
+            }
+        );
+    }
+
+    // Cross validate the length and md5 for a pair of sequence records and throw if they're not the same.
+    private void validateSequenceDictionaryRecords(
+            final String sourceDataSourceName,
+            final SAMSequenceDictionary sourceDictionary,
+            final SAMSequenceRecord sourceSequence,
+            final String targetDataSourceName,
+            final SAMSequenceDictionary targetDictionary,
+            final SAMSequenceRecord targetSequence)
+    {
+        if (targetSequence.getSequenceLength() != sourceSequence.getSequenceLength()) {
+            final String msg = String.format("Incompatible sequence lengths found (%s: %d) and (%s: %d)",
+                    sourceSequence.getSequenceName(),
+                    sourceSequence.getSequenceLength(),
+                    targetSequence.getSequenceName(),
+                    targetSequence.getSequenceLength());
+            throw new UserException.IncompatibleSequenceDictionaries(
+                    msg,
+                    sourceDataSourceName,
+                    sourceDictionary,
+                    targetDataSourceName,
+                    targetDictionary
+            );
+        } else if (!Objects.equals(targetSequence.getMd5(), sourceSequence.getMd5())) {
+            final String msg = String.format("Incompatible sequence MD5 values found (%s: %s) and (%s: %s)",
+                    sourceSequence.getSequenceName(),
+                    sourceSequence.getMd5(),
+                    targetSequence.getSequenceName(),
+                    targetSequence.getMd5());
+            throw new UserException.IncompatibleSequenceDictionaries(
+                    msg,
+                    sourceDataSourceName,
+                    sourceDictionary,
+                    targetDataSourceName,
+                    targetDictionary
+            );
+        }
+    }
+
+    /**
+     * Close the iterator currently open over the data sources, if there is one.
+     */
+    private void closeOpenIterationIfNecessary() {
```

Comment:
I think your `close()` method should be calling `closeOpenIterationIfNecessary()`